ENH: Implement DOP853 ODE integrator

nmayorov opened the pull request:
I tried to figure out DOP853 algorithm in order to finish #9290 

Eventually I decided to make a `RungeKutta` subclass implementation trying to make it more straightforward.

The main features of DOP853 compared to RK45 and RK23:

1. There are two error estimators. Both estimators are implemented in the same way --- by a linear combination of computed RK stages. Then the final error is computed by a certain combination. This requires very little modification of the original code, two sets of coefficients needs to be defined.
2. Three extra stages are required to compute the interpolant. This again is relatively straightforward and done inside `_dense_output_impl`.
3. The suggested interpolant is not expressed in the polynomial basis. Of course it can be expressed as such, but to use the original coefficients we need to use the suggested evaluation scheme. This is implemented inside `Dop853DenseOutput`.
4. Enhanced step size control is suggested with parameter `beta`. I haven't implemented this yet and put it as a low priority.


----------------------------------

### Comparison of RK23, RK45 and DOP853 on a celestial mechanics problem:

```Python
import numpy as np
from scipy.integrate import solve_ivp


mu = 0.0122771
nu = 1 - mu

t0 = 0
tf = 17.0652165601579625588917206249
y0 = [0.994, 0, 0, -2.00158510637908252240537862224]

rtol=1e-10
atol=1e-14


def fun(x, y):
    D1 = ((y[0] + mu)**2 + y[1]**2) ** 1.5
    D2 = ((y[0] - nu)**2 + y[1]**2) ** 1.5

    return np.array([
        y[2],
        y[3],
        y[0] + 2 * y[3] - nu * (y[0] + mu) / D1 - mu * (y[0] - nu) / D2,
        y[1] - 2 * y[2] - nu * y[1] / D1 - mu * y[1] / D2
    ])


np.set_printoptions(10)
for method in ['RK23', 'RK45', 'DOP853']:
    res = solve_ivp(fun, [t0, tf], y0, method=method, rtol=rtol, atol=atol)
    print(method)
    print("Solution at tf: {}".format(res.y[:, -1]))
    print("Number of function evaluations: {}".format(res.nfev))
    print("--------------------------------------------------")
```

The result:
```
RK23
Solution at tf: [ 0.9929909741 -0.0024338004 -0.4547767983 -2.0295309619]
Number of function evaluations: 99674
--------------------------------------------------
RK45
Solution at tf: [ 0.9929909646 -0.0024338146 -0.4547799205 -2.0295307335]
Number of function evaluations: 7622
--------------------------------------------------
DOP853
Solution at tf: [ 0.9929909675 -0.0024338098 -0.4547789079 -2.0295308438]
Number of function evaluations: 4046
--------------------------------------------------
```

Dense output illustration with `rtol=1e-8` (same `atol`):
![traj](https://user-images.githubusercontent.com/7074023/57582014-72370380-74c8-11e9-9016-d9dfb38f55b5.png)

Overall it seems to work correctly and perform more efficient with low tolerances.

However currently it does ~20% more function evaluations compared to the Fortran code. It must have something to do with error estimation and step size control / step rejection. At least I want to determine the reason for it. But I certainly don't plan to make it behave exactly as the Fortran code --- hard to achieve, creates strange code and doesn't give any real benefits.

---------

@Juanlu001 @nikita-astronaut 

I hope you guys are still interested in making it part of scipy and don't mind that I took a different approach to the implementation. Your code was certainly useful for me.

Juanlu001 commented:
Hi @nmayorov, thanks a lot for taking this over! I'm so happy that it's on good hands :) And glad my original attempt was helpful!

Regarding the number of function evaluations:

> At least I want to determine the reason for it. But I certainly don't plan to make it behave exactly as the Fortran code --- hard to achieve, creates strange code and doesn't give any real benefits.

I was thinking of trying a side-by-side debugging, but perhaps that's not the best strategy if the code is not going to be exactly the same. Still, I will take a look at it this week.

On the other hand, we will try this method in [poliastro](https://github.com/poliastro/poliastro/) and report here how it went.

Juanlu001 commented:
Pinging @helgee as well in case he wants to take a look.

jorgepiloto commented:
@nmayorov This is amazing! Thank you for this contribution! As @Juanlu001 said we will be testing this in [poliastro](https://github.com/poliastro/poliastro) :rocket: 

I will take a closer look to your implementation for DOP853 and study it in depth since we are also working on Verner78 method. Some  time ago I opened an issue #9698 because I got stacked with the implementation of this method from Runge-Kutta class. This is the official page for [Verner's Methods](http://people.math.sfu.ca/~jverner/), from which we took the coefficients.

Again, thank you for this pull request! :smile:

nmayorov commented:
Hey guys! Good news --- I have figured out what the difference between this code and Fortran code is and will make an update soon. 

It comes down to 2 main things:

1. Error estimation is done somewhat different to what I implemented. The denominator `(err5**2 + err3**2 * 0.01) ** 0.5` is computed not component-wise but as a scalar using norms of the corresponding errors (with `scale` involved).
2. After a step was rejected the original code does not allow for the step size to be increased after a single successful step.

After I implemented these changes I got the `nfev` higher by exactly the number of rejected steps. Why is that? Because I always compute `f_new = fun(t_new, y_new)` (for generality), but DOP853 requires this only when the step was accepted (i.e. the error estimation does not require `f_new`).

My suggestion is not to change this place in order not to complicate the code.

nmayorov commented:
The updated result on the same problem:

```
RK23
Solution at tf: [ 0.9929909729 -0.0024338022 -0.4547771854 -2.0295309264]
Number of function evaluations: 100658
--------------------------------------------------
RK45
Solution at tf: [ 0.9929909649 -0.0024338142 -0.4547798365 -2.0295307436]
Number of function evaluations: 7094
--------------------------------------------------
DOP853
Solution at tf: [ 0.9929909685 -0.0024338082 -0.4547785701 -2.0295308818]
Number of function evaluations: 3506
--------------------------------------------------
```
The remaining difference of the implementation:

1. Minimum and maximum value for step size change factors. I don't think these are terribly important, probably we can keep the current values.
2. Initial step size selection. It seems to make a bigger difference, I will study what is implemented in DOP853 Fortran code and perhaps transfer this implementation to Python. I would prefer to have a single implementation of the step size selection (at least for RK methods).

nmayorov commented:
Re-evaluated the situation quite a bit. I did compare the results to the MATLAB version posted on the same page with Fortran code.

And considering MATLAB version --- the codes work similar, the nfev is higher by the number of rejected steps which I explained above. However, the Fortran code (called through scipy) still works a bit different. But this code is painful to study in microscopic details.

As for the initial step selection --- original code uses norms without division by the number of components, which I believe was some arbitrary decision by the authors (perhaps small "mistake" on their part, at least it seems to contradict the description from the book). But I will think about it some more.

-----------------

My current perspective is to keep the current version of the code as a legit implementation of DOP853. It looks conceptually correct, has right properties and works very similar to the MATLAB version and quite similar to the original Fortran version.

Please report the result of your evaluation within polyastro.

nikita-astronaut commented:
Hi @nmayorov , 

thanks a lot for pushing it through!

Juanlu001 commented:
All our tests passed, and I did some quick tests and I'd say that this implementation is noticeably faster than ours, although I didn't measure it. Looking forward to see it merged, too bad we'll have to wait for SciPy 1.4 :)

ApamNapat commented on scipy/integrate/_ivp/common.py:
```diff
@@ -80,7 +80,8 @@ def select_initial_step(fun, t0, y0, f0, direction, order, rtol, atol):
     order : float
-        Method order.
+        Error estimator order. It means that the error controlled by the
+        algorithm is proportional to ``step_size ** (order + 1)`.
```
I think you have one ` too many or you're missing one

nmayorov replied:
It is a little confusing, but I think it is correct. In numerical ODE solving an algorithm order is meant as its global error (under some "good conditions") and thus the local error is proportional to `h ** (order + 1)`.

nmayorov commented:
@rgommers 

Not sure how to proceed with it. Could you suggest something please? I think we can merge it.

rgommers commented:
It looks like @Juanlu001 and @nikita-astronaut are happy with this, so I think we should get this merged indeed. Let's see if CI is completely happy. I'm not much of an expert, but happy to have a look over this PR this week and merge it (if anyone wants to beat me to it, please feel free!).

nmayorov commented:
@rgommers sounds great, thanks.

tylerjereddy commented on scipy/integrate/_ivp/rk.py:
```diff
@@ -3,7 +3,7 @@
 from .common import (validate_max_step, validate_tol, select_initial_step,
                      norm, warn_extraneous, validate_first_step)
-
+from . import dop853_coefficients
```
Just a question from a design standpoint--it looks like `RK45` & `RK23` are also in this module but don't use these coefficients? Since the import executes some top-level code in `dop853_coefficients` to intialize a bunch of values, is there something to be said for confining these operations to only happen within the `DOP853` class initialization?

nmayorov replied:
It seems quite hard to control such things. All these classes are imported in `ivp.py`, so I guess these coefficients are going into memory anyway. Doing some "lazy" initialization --- I think it creates unnecessary complexity and does more harm than good. After all the size of the coefficients arrays is negligible.

Another way to organize the code is to put `DOP853` and all the coefficients within one file. But there are some common constants for Runge-Kutta methods, so it's also not ideal.

rgommers replied:
agreed, this is very low-cost so should be fine to keep as is

tylerjereddy commented on scipy/integrate/_ivp/tests/test_ivp.py:
```diff
@@ -665,7 +676,7 @@ def test_ConstantDenseOutput():
 def test_classes():
     y0 = [1 / 3, 2 / 9]
-    for cls in [RK23, RK45, Radau, BDF, LSODA]:
+    for cls in [RK23, RK45, DOP853, Radau, BDF, LSODA]:
```
You parametrize the new test file nicely below, but actually not parametrizing here might the right choice for now--helps keep the diff clean of extra refactoring for a PR that is already quite large.

nmayorov replied:
I guess this more logically fits into another pull request indeed.

tylerjereddy commented on scipy/integrate/_ivp/rk.py:
```diff
@@ -364,6 +383,147 @@ class RK45(RungeKutta):
+        relative accuracy (number of correct digits). But if a component of `y`
+        is approximately below `atol`, the error only needs to fall within
+        the same `atol` threshold, and the number of correct digits is not
+        guaranteed. If components of y have different scales, it might be
```
`y` is in single backticks above, so maybe here too?

nmayorov replied:
Agree with it. I started making the change everywhere but then noticed that it creates too much unrelated diff. I think it's better to be fixed in all places in a separate PR.

rgommers replied:
as long as doc build CI passes (which it does here), I think it's fine to leave this for a separate cleanup PR indeed.

rgommers commented:
@nmayorov if you'd like to add a release note in https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0, that would be useful

Juanlu001 commented:
Thanks @nmayorov and everyone! We're so happy to see this in SciPy ♥️

nmayorov commented:
@rgommers updated release notes about DOP853.

@Juanlu001 happy to be useful!
